### PR TITLE
feat: dependent features

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

**Spec for dependent features**

Data structure for dependent features:
* depend on parent being enabled: `{feature: 'parent'}`
* depend on parent being disabled (useful for kill-switches): `{feature: 'parent', enabled: false}`
* depend on parent resolving to a value/variant: `{feature: 'parent', variants: ['parentVariant', 'anotherParentVariant']}`

We're deliberately choosing `feature` as a key to have an option for other types of dependencies e.g. strategies

Most of the code has the same activation strategy with a strategy variant. Enabled tests and variant tests are almost identical (only description and expected result are different).

Interesting cases:
* you can depend on many features (with AND operator between them)
* when you depend on a parent feature variant you can specify a list of options that you'd like to match against (OR operator between variants)
* we don't support transitive dependencies, only direct dependencies

Reference implementation passing spec tests: https://github.com/Unleash/unleash-client-node/pull/520

### CODE REVIEW

File: specifications/17-dependent-features.json
Comment (sighphyre) [inline specifications/17-dependent-features.json:49]: Oh cool, new fear unlocked!

What's the actual data structure enforced by the server? Is this a graph or a tree? 

If it's a graph then cycle detection is much harder than this in the case where you have A -> B -> C -> A (please say it's not a graph, or if it is, a DAG enforced by the server)

If it's a tree, this shouldn't happen (but it's nice to be defensive!), hopefully that can be checked and blocked by the server

Comment (kwasniew) [reply specifications/17-dependent-features.json:49]: There are no enforcements in the backend yet. But the rule will be simple: if your parent has any dependency it can't be your parent. This is our iteration 0.

Comment (kwasniew) [reply specifications/17-dependent-features.json:49]: This is the line in SDK: 
```ts
if (this.repository.getToggle(parent.feature)?.dependencies?.length) {
        return false;
}
```

File: specifications/17-dependent-features.json
Comment (sighphyre) [inline specifications/17-dependent-features.json:558]: Nice, that's what I would have done too, I think this is correct. This is a big decision though. It means stale caches and/or cycling out your api key can potentially change the state of a toggle. Leaving this here for visibility in case someone has a different opinion

Comment (kwasniew) [reply specifications/17-dependent-features.json:558]: Glad you're confirming my confirmation bias :) As you said it's up for a discussion.

**Review by sighphyre — COMMENTED** [review]
I think it's worth including some tests that also depend on context properties, e.g. a constraint. Based on the node implementation the desired behaviour appears to be that the context is reused for the parent check as well. Worth enforcing here I think

File: specifications/17-dependent-features.json
Comment (sighphyre) [reply specifications/17-dependent-features.json:49]: > There are no enforcements in the backend yet. But the rule will be simple: if your parent has any dependency it can't be your parent. This is our iteration 0.

Awesome, that should handle cycles in either data structure. I think it's worth leaving in, but the only way this can trigger then is if you play silly buggers with your backup files

**Comment by kwasniew** [comment]
@sighphyre addressed all your points :)

**Review by sighphyre — APPROVED** [review]
What a legend. This looks rather nice

I think others should weigh in but LGTM

File: specifications/17-dependent-features.json
Comment (FredrikOseberg) [reply specifications/17-dependent-features.json:663]: This is tricky. Since you can filter by tags it's technically possible to end up in a situation where the parent is not present. We'll need to implement safeguards in the SDK to warn you if this happens, but I think the behavior here is correct. If you depend on a feature toggle that you have not fetched, there is no way for us to know the evaluation result of that toggle and we can not determine that the child should be enabled.

Comment (FredrikOseberg) [reply specifications/17-dependent-features.json:663]: I guess this is where a feature toggle type would come in handy @kwasniew

Comment (kwasniew) [reply specifications/17-dependent-features.json:663]: @FredrikOseberg I will try to add a warning when you try to depend on a feature that you don't have in the SDK

Comment (kwasniew) [reply specifications/17-dependent-features.json:663]: @FredrikOseberg can you elaborate on a feature toggle type?

Comment (FredrikOseberg) [reply specifications/17-dependent-features.json:663]: Yes. Like we talked about in the meeting, some way to mark a toggle as dependable, we discussed the possibility of having a new feature toggle type called Dependable Toggle. If we had that, we could circumvent the filtering in the API and say that "No, we don't care that you are filtering by tags, you'll get all the dependable toggles in this project regardless".

File: specifications/17-dependent-features.json
Comment (daveleek) [inline specifications/17-dependent-features.json:454]: This operates on `results` I take it? Do default-variants count as []?

File: specifications/17-dependent-features.json
Comment (kwasniew) [reply specifications/17-dependent-features.json:663]: @FredrikOseberg added reference impl for warnings: https://github.com/Unleash/unleash-client-node/pull/520/commits/1ed0994bb2c598e8289320d3a0b4243a90e4a785

Comment (kwasniew) [reply specifications/17-dependent-features.json:663]: Now I remember the dependent feature markers we talked about :) We definitely need to keep it in mind. For this spec though I think we can proceed without this concept since it will only affect the client API not this spec.

Comment (FredrikOseberg) [reply specifications/17-dependent-features.json:663]: Agree. This spec just surfaced the problem, so I just mentioned it here.

File: specifications/17-dependent-features.json
Comment (daveleek) [reply specifications/17-dependent-features.json:454]: Ah, great! So `"variants": [ ]` is the same as not including the variants property in the dependency

**Review by daveleek — APPROVED** [review]
I like it!

**Pull request merged**
